Remove dropped table directory

bhalevy opened the pull request:
This series adds removal of dropped table directory when it has no remaining snapshots.

There are 2 code paths that take of that:
1. when the table is dropped and there are no active snapshots for it (typically when auto_snapshot disabled).
2. or when the last snapshot is cleared, leaving no other snapshot for a dropped table.

Unit tests were extended to covert these scenarios.

Fixes #10896

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1252/

nyh commented on test/boost/database_test.cc:
```diff
@@ -1101,3 +1101,48 @@ SEASTAR_TEST_CASE(database_drop_column_family_clears_querier_cache) {
+    });
+
+    // Disallow auto_snapshot
+    // dir should not exist
```
I don't know what I'm missing, but it seems like you just have the same test three times, and it doesn't "allow" or "disallow" auto_snapshot anywhere?
I guess you'll go back to fix this in a later patch?

bhalevy replied:
`auto_snapshot` is allowed by default in the cql_test_env db::config.
The last boolean arg to `drop_table_on_all` is `with_snapshot`, that, when true, allows taking a snapshot, based on the `auto_snapshot` configuration, and when false, skips the snapshot.  We pass `true` the first time and `false` the second time.  As for the auto_snapshot config, this is done in a later patch.  Maybe I'll pull some of the logic here so it would make more sense.

BTW, I want to eventually change this boolean parameter to a tri-state class: `{auto, yes, no}` so its usage will be more visible, with the following semantics: auto: consider the auto_snapshot config, yes: take a snapshot, no: don't take a snapshot.  But it is out of scope for this series and will make it harder to backport.

nyh commented on test/boost/database_test.cc:
```diff
@@ -1145,5 +1150,10 @@ SEASTAR_TEST_CASE(drop_table_with_snapshots) {
         BOOST_REQUIRE_EQUAL(cf_dir_exists, true);
         co_return;
-    });
+    }, db_cfg_ptr);
```
I still don't understand why this function has two (even three?) copies of the same code?

nyh replied:
I think I understand now - the three copies do different things, but only the first one uses the "auto_snapshot" parameter as far as I can see, so maybe they could/should have been separate functions.

nyh commented on test/boost/database_test.cc:
```diff
@@ -575,21 +575,56 @@ SEASTAR_TEST_CASE(snapshot_list_inexistent) {
-            return make_ready_future<>();
-        }).get();
-        BOOST_REQUIRE_GT(count, 1); // expect more than the manifest alone
+    auto snapshot_name = [] (int idx) {
```
This test checks a lot of details (like the snapshot's name) that the previous test didn't check. So I wonder if it didn't make sense to keep both tests - the small one and the larger one? But maybe not. Your call.

bhalevy replied:
Hmm, not sure either.  Maybe it'd be best to just keep the simple test since the extended test is not a pure superset of the simple test (i.e. it doesn't test exactly the same condition but does more)

nyh commented on replica/database.cc:
```diff
@@ -2432,15 +2432,15 @@ static sstring get_snapshot_table_dir_prefix(const sstring& table_name) {
+    constexpr size_t uuid_size = 32;
+    ssize_t pos = directory_name.size() - uuid_size - 1;
+    if (pos <= 0 || directory_name[pos] != '-') {
+        on_internal_error(dblog, format("table directory entry name '{}' is invalid: no '-' separator found at pos {}", directory_name, pos));
```
I want to verify something: Is there some API where this directory_name comes from the user input? If there is, it shouldn't be an on_internal_error but just a regular exception.

bhalevy replied:
No, the function is never called directly outside of the database layer,
and there it is called for "sanitized" directory entries that are listed in keyspace directories residing under the configured data directories.

nyh commented on replica/database.cc:
```diff
@@ -2571,21 +2571,37 @@ future<> database::clear_snapshot(sstring tag, std::vector<sstring> keyspace_nam
+                        auto dropped = (it == _ks_cf_to_uuid.cend()) || (cf_uuid != it->second);
+                        if (dropped) {
+                            dblog.info("Removing dropped table dir {}", table_dir);
+                            recursive_remove_directory(table_dir).get();
```
I can't find any bug in this code, but it looks scary - if the logic above is somewhat wrong, you can find yourself removing an entire non-empty directory. If table_dir is somehow (?) "/", you find yourself removing the entire disk ;-)
Again, judicious use of unlink() to remove only guaranteed-empty directories seems safer to me, although I admit there's nothing wrong in your code that I can find.

bhalevy replied:
No problem.  I'll give the safer unlink approach a shot.
Although I will use `recursive_remove_directory` for table subdirectories, but this is safe than removing just `table_dir`.
It should never ever become "/" or empty, but I guess a bug in the future might cause that, and adding a subdirectory to it will prevent/limit the damage.

nyh replied:
I thought (but maybe I'm missing something) that you can assume that at this stage, there would only be empty subdirectories - there will be an empty "staging" subdirectory (because you said it's removed earlier), an empty "snapshots" subdirectory (if there are no snapshots), etc. So you can use unlink() for all of them. If unlink("snapshots") fails, it means there are snapshots, and you don't even need to check.
What subdirectory can be non-empty and require recursive_remove_directory?

bhalevy replied:
Normally, the sub-directories would be empty, but I really don't think we should care so much about it.
Otherwise, we won't be able to get rid of stale directories of dropped tables that happened to have some unexpected junk in them (from the point of view of a pedantic function).  For example, a table that was removed using an older version of scylla, or a table we want to drop after an error (e.g. ENOSPC) that might have stuff under the "pending_delete" subdirectory, that can't proceed for some reason.

Bottom line, is I think that making this function pedantic might require too much code and I don't see the benefit in that.
If the user asked to drop the table we should remove everything under the table directory other than live snapshots.

We may need special care for snapshots in the future though since I'm considering a change in the snapshot taking logic to create the snapshot in a temporary directory and move it to its final name only when taking the snapshot is successfully done, to prevent cases where we end up with partially built snapshot after an error.  Those temporary snapshot directories should be cleaned up after the error, and we can clean them up on startup, but I think we should also delete them unconditionally when removing the table dir.

nyh reviewed (approved):
I'm approving, but I left several comments that I want to ask you to consider.
One thing (maybe I misunderstood?) it seems one of your tests does the same thing two or three times.
Another thing is that you seem to have gone into a lot of effort to check when there are or aren't snapshots, and I wonder if everything wouldn't have just been simpler if you used unlink() to remove empty directories - and do nothing if they are not empty.  Although it seems your code is correct too.

bhalevy commented on replica/database.cc:
```diff
@@ -1029,11 +1030,29 @@ future<> database::drop_column_family(const sstring& ks_name, const sstring& cf_
+        co_return;
+    }
+    try {
+        co_await recursive_remove_directory(cf_path);
```
There is also a potential quarantine subdirectory too, but I think that approach should work.

Regarding the staging directory, it should be empty as we don't allow dropping tables that have views, or we drop them automatically from the alternator path, and that should happen before the base table is dropped.

Note: Need to augment documentation https://docs.scylladb.com/getting-started/ddl/#drop-table-statement about that,

bhalevy replied:
https://github.com/scylladb/scylla-docs/pull/4129

avikivity replied:
unlink = stall, must use seastar APIs

bhalevy replied:
> unlink = stall, must use seastar APIs

yes, this was only suggested by @nyh, but never made it to any version of this PR branch.
the equivalent call to `unlink` is `seastar::remove_file`.

nyh replied:
Obviously, when I suggested unlink() I didn't mean we should use the Linux system call directly.

bhalevy commented:
In v2 (f95150ccc5):
- Added:
  - "sstables: officially define pending_delete_dir"
  - "sstables: define table_subdirectories"
- moved dir removal code to `sstables::remove_table_directory_if_has_no_snapshots`
  - and make it safer
  - use in `drop_table_on_all` and `clear_snapshot`
- improved comments in `database_test.drop_table_with_snapshots`
- Added "database_test: add clear_multiple_snapshots test"
  - rather than  extending the existing "clear_snapshot" test case

nyh commented on sstables/sstables.hh:
```diff
@@ -123,6 +123,14 @@ constexpr const char* snapshots_dir = "snapshots";
 constexpr const char* quarantine_dir = "quarantine";
 constexpr const char* pending_delete_dir = "pending_delete";
 
+constexpr std::array<const char*, 5> table_subdirectories = {
```
If the "5" here bothers you (like it did me when I saw it ;-)), you can also use std::to_array - see https://en.cppreference.com/w/cpp/container/array/to_array. It's new in C++20.

avikivity replied:
Plain std::array without the template parameters will work just as well.

avikivity replied:
But, better to move it out of the header.

nyh commented on sstables/sstables.cc:
```diff
@@ -3320,8 +3320,63 @@ gc_clock::time_point sstable::get_gc_before_for_fully_expire(const gc_clock::tim
+    }
+
+    // Return error code, 0 is success
+    auto remove_dir = [] (fs::path dir, bool recursive) -> future<int> {
```
Nitpick: why make this a lambda instead of a normal static function?

bhalevy replied:
I'm reluctant to make it available to other users since it's quite specific to the calling function,
but we can make it a private static function at this point until we find other users for it.

nyh replied:
Unless I'm missing something, this simple function doesn't use anything from the class or anything. It can be an old-school "static" function (a file-local function). It doesn't need to be a "private static" function in the C++ classes sense.
But I don't insist. If you like this lambda thing, feel free to keep it.

bhalevy replied:
true, I'll make it a static free function then.
it may help readability of the code.

nyh commented on sstables/sstables.cc:
```diff
@@ -3320,8 +3320,63 @@ gc_clock::time_point sstable::get_gc_before_for_fully_expire(const gc_clock::tim
+    int error = 0;
+    for (auto subdir : sstables::table_subdirectories) {
+        // Remove the snapshot directory only if empty
+        // while other subdirectories are removed recusresively.
```
I'm still curious why remove the other subdirectories recursively. What can be there? Shouldn't the other stuff be removed when you stop the view building, stop the repair, and so on - and we expect all of them to be empty too?

bhalevy replied:
We do expect them to be empty on the error-free path.
I just don't want to depend on that when removing the table, since it will make the remove path more fragile, and needlessly so.

nyh commented on sstables/sstables.cc:
```diff
@@ -3320,8 +3320,63 @@ gc_clock::time_point sstable::get_gc_before_for_fully_expire(const gc_clock::tim
+        // Remove the table directory recursively
+        // since it may still hold leftover temporary
+        // sstable files and directories
+        co_await remove_dir(table_dir, true);
```
Oh, I didn't realize that this can happen when I suggested the unlink() method. Can it happen? Why would there be any "temporary sstable files"? Don't we delete those after a reboot?

avikivity commented:
subject: s/dir$directory/

avikivity commented on replica/database.cc:
```diff
@@ -1028,6 +1028,13 @@ future<> database::drop_column_family(const sstring& ks_name, const sstring& cf_
     f.get(); // re-throw exception from truncate() if any
 }
 
+future<> database::drop_table_on_all(sstring ks_name, sstring cf_name, timestamp_func tsf, bool with_snapshot) {
```
A better name is drop_table_on_all_shards(), so people don't have to wonder on all what (invoke_on_all is a sharded<> method, so that provides some context).

Another way to go is static database::drop_table_on_all_shards(sharded<database>&), which makes it even clearer by the type of the first parameter.

bhalevy replied:
sure, will do.

bhalevy replied:
Using becomes quite long and cumbersome, e.g.
`replica::database::drop_table_on_all_shards(e.db(), etc...)`

How about shortening it to `replica::database::drop_table(e.db(), etc...)`
The caller shouldn't care how it's done internally in the `sharded<database>`, does it matter to the caller if we decided to implement it e.g. only on shard 0, or on the local shard?

bhalevy replied:
Changed to a static `drop_table_on_all_shards` method in v4.

avikivity replied:
Long it better than short when things are not as expected (operating on the sharded container rather than the object called on)

avikivity commented on sstables/sstables.hh:
```diff
@@ -123,6 +123,14 @@ constexpr const char* snapshots_dir = "snapshots";
+    quarantine_dir,
+    pending_delete_dir,
+});
+
```
I made a comment about moving it out-of-line, but since the constituents are still inline, you can keep it like that for now and we'll clean everything up later.

avikivity commented on sstables/sstables.cc:
```diff
@@ -3320,8 +3320,63 @@ gc_clock::time_point sstable::get_gc_before_for_fully_expire(const gc_clock::tim
+        int error_code = -1;
+        try {
+            std::rethrow_exception(ex);
+        } catch (const std::system_error& e) {
```
Minor, but you can just catch std::system_error in the outer catch.

bhalevy replied:
Done in v4

avikivity commented on db/schema_tables.cc:
```diff
@@ -1289,19 +1289,19 @@ static future<> merge_tables_and_views(distributed<service::storage_proxy>& prox
+    });
+    co_await coroutine::parallel_for_each(tables_diff.dropped, [&db = proxy.local().get_db().local()] (schema_diff::dropped_schema& dt) -> future<> {
+        auto& s = *dt.schema.get();
+        return db.drop_table_on_all(s.ks_name(), s.cf_name(), [&] { return dt.jp.value(); });
```
Should we make database::drop_column_family private?

bhalevy replied:
It is, in the next patch.

avikivity commented on replica/database.hh:
```diff
@@ -1646,8 +1646,9 @@ public:
+private:
     future<> drop_column_family(const sstring& ks_name, const sstring& cf_name, timestamp_func, bool with_snapshot = true);
-
+public:
```
yes

avikivity commented:
Looks good, minor comments.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1283/

bhalevy commented:
In v4 (bbbbea65fb)
- rebased
- database: add drop_table_on_all_shards
  - renamed, made static
- database: drop_table_on_all_shards: remove table directory having no snapshots
  - catch (std::system_error&) in outer try/catch block.
  - split the long unit test that is added to multiple smaller ones
- database_test: drop_table_with_snapshots: test auto_snapshot
  - focus only on the `drop_table_with_auto_snapshot` test case

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1389/

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1395/
